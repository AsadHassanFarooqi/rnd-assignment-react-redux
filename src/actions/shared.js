import { setAuthedUser } from "./authedUser";
import { receiveUsers } from "./users";
import { receiveQuestions } from "./questions";
import { getInitialData } from "../utils/api";

import Cat from '../images/cat.png';

const user = {
    sarahedo: {
        id: 'sarahedo',
        password:'password123',
        name: 'Sarah Edo',
        avatarURL: Cat,
        answers: {
          "8xf0y6ziyjabvozdd253nd": 'optionOne',
          "6ni6ok3ym7mf1p33lnez": 'optionOne',
          "am8ehyc8byjqgar0jgpub9": 'optionTwo',
          "loxhs1bqm25b708cmbf3g": 'optionTwo'
        },
        questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
      },
}

export function handleInitialData() {
    return (dispatch) => {
        return getInitialData().then(({users, questions}) => {
            dispatch(receiveUsers(users));
            dispatch(receiveQuestions(questions));
        });
    };
}
