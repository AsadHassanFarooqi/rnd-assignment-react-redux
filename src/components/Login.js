import React, { useState } from 'react'
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Image from 'react-bootstrap/Image'
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import { connect } from 'react-redux';
import { useNavigate } from "react-router-dom";

import { handleLogin } from "../actions/authedUser";

import pollsImage from '../images/polls-image.png';

const Login = ({dispatch, users}) => {

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const navigate = useNavigate();

  const handleUsername = (e) => {
    const userID = e.target.value;
    setUsername(userID);
  }

  const handlePassword = (e) => {
    const userPassword = e.target.value;
    setPassword(userPassword);
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(handleLogin(username, password));
    setUsername('');
    setPassword('');
    navigate(-1);
  }

  return (
    <Row className="mx-auto d-flex justify-content-center align-items-center">
      <Col xs="12 mt-5"><h1 className="text-center">Employee Polls Project</h1></Col>
      <Col xs="12" lg="8">
        <Image src={pollsImage} alt="Polls" className="mx-auto mt-2" fluid rounded />
      </Col>
      <Col lg="8">
        <Form className="mt-5" onSubmit={handleSubmit}>
          <Form.Group className="mb-3">
            <Form.Label>User ID</Form.Label>
            <Form.Select aria-label="users select" value={username} onChange={handleUsername} data-testid="user-field">
              <option>Open this select user</option>
              {Object.values(users).map(user => (
                <option value={user.id} key={user.id}>{user.id}</option>
              ))}
            </Form.Select>
          </Form.Group>

          <Form.Group className="mb-3">
            <Form.Label>Password</Form.Label>
            <Form.Control type="password" placeholder="Password" value={password} name="password" onChange={handlePassword} data-testid="password-field" />
          </Form.Group>
          <div className="text-center">
            <Button variant="primary" type="submit" data-testid="login-btn">
              Submit
            </Button>
          </div>
        </Form>
      </Col>
    </Row>
  )
}

const mapStateToProps = ({users}) => ({
  users
})

export default connect(mapStateToProps)(Login);