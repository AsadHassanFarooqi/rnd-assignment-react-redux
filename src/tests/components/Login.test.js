import React from "react";
import {Provider} from "react-redux";
import {fireEvent, render} from "@testing-library/react";
import {BrowserRouter} from "react-router-dom";

import {store} from "../../store";
import Login from "../../components/Login";
import { handleLogin } from '../../actions/authedUser';

describe("Login", () => {
  it("should render the component", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login/>
        </BrowserRouter>
      </Provider>
    );
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  it("will verify that the login component has a login field, a password field and a submit button", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    const userInput = view.getByTestId("user-field");
    const passwordInput = view.getByTestId("password-field");
    const loginButton = view.getByTestId("login-btn");
    
    expect(userInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(loginButton).toBeInTheDocument();

    fireEvent.change(userInput, { target: { value: "sarahedo" } });
    fireEvent.change(passwordInput, { target: { value: "password123" } });
    fireEvent.click(loginButton);
    store.dispatch(handleLogin(userInput.value, passwordInput.value));
  });
});
