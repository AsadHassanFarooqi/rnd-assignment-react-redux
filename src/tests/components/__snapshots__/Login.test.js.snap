// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mx-auto d-flex justify-content-center align-items-center row"
      >
        <div
          class="col-12 mt-5"
        >
          <h1
            class="text-center"
          >
            Employee Polls Project
          </h1>
        </div>
        <div
          class="col-lg-8 col-12"
        >
          <img
            alt="Polls"
            class="mx-auto mt-2 img-fluid rounded"
            src="polls-image.png"
          />
        </div>
        <div
          class="col-lg-8"
        >
          <form
            class="mt-5"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
              >
                User ID
              </label>
              <select
                aria-label="users select"
                class="form-select"
                data-testid="user-field"
              >
                <option>
                  Open this select user
                </option>
              </select>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
              >
                Password
              </label>
              <input
                class="form-control"
                data-testid="password-field"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="text-center"
            >
              <button
                class="btn btn-primary"
                data-testid="login-btn"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mx-auto d-flex justify-content-center align-items-center row"
    >
      <div
        class="col-12 mt-5"
      >
        <h1
          class="text-center"
        >
          Employee Polls Project
        </h1>
      </div>
      <div
        class="col-lg-8 col-12"
      >
        <img
          alt="Polls"
          class="mx-auto mt-2 img-fluid rounded"
          src="polls-image.png"
        />
      </div>
      <div
        class="col-lg-8"
      >
        <form
          class="mt-5"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              User ID
            </label>
            <select
              aria-label="users select"
              class="form-select"
              data-testid="user-field"
            >
              <option>
                Open this select user
              </option>
            </select>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="password-field"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="text-center"
          >
            <button
              class="btn btn-primary"
              data-testid="login-btn"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
